status-handler.js

```diff
@@ -172,19 +177,45 @@ class StatusHandler {
    * Check if a build is in a queued status
    *
    * @function _isBuildQueued
-   * @param {Object} build - Build params
-   * @param {String} build.pkg - The package name
-   * @param {String} build.version - The package version
-   * @param {String} build.env - The build environnement
+   * @param {Object} data - Message from NSQ
    * @returns {String} the package name
    */
-  async _isBuildQueued(build) {
+  async _isBuildQueued(data) {
+    const { version, env } = data;
+    const pkg = this._getPackageName(data);
+
     // Fetch list of previous events for a certain build
-    const previousEvents = await this.models.StatusEvent.findAll(build);
+    const previousEvents = await this.models.StatusEvent.findAll({
+      pkg, version, env
+    });
+
+    const nEvents = previousEvents.length;
+
+    if (nEvents === 0) return false;
+
+    const lastEvent = previousEvents[nEvents - 1];
+
+    // Ensure that second last event is 'Builds Queued' from carpentd
+    // and that current event is 'Fetched tarball' from carpentd-worker
+    if (
+      lastEvent.message === MSG_FETCHED_TARBALL &&
+      lastEvent.locale === data.locale &&
+      data.message === MSG_FETCHED_TARBALL &&
+      previousEvents[nEvents - 2].message === MSG_BUILD_QUEUED &&
+      nEvents > 1
+    ) {
+      return true;
+    }
+
+    // Same as before but handle if StatusEvent.findAll does not return
+    // current event written in the database yet becaue of eventual consistency
+    if (
+      lastEvent.message === MSG_BUILD_QUEUED &&
+      data.message === MSG_FETCHED_TARBALL
+    ) {
+      return true;
+    }
```

Comment:
There's only 3 of these so far and its already feeling a bit unwieldy can we find a way to organize these better so that it's not going to go crazy as we add more things we want to webhook on?

Reply:
This is the logic for `_isBuildQueued`. Conditions won't adds up here. We may going to have some other logic for other web hooks type in the future but it will be in a separate function IF AND ONLY IF needed. In that case, I will also consider possible refactor.